WcaOnRails/lib/results_validators/individual_results_validator.rb

```diff
@@ -0,0 +1,310 @@
+      unless sum_of_times_for_rounds < time_limit_for_round.centiseconds
+        @errors << ValidationError.new(:results, competition_id,
+                                       RESULTS_OVER_CUMULATIVE_TIME_LIMIT_ERROR,
+                                       round_ids: cumulative_round_ids.join(","),
+                                       person_name: person_name,
+                                       time_limit: time_limit_for_round.to_s(round_info))
+      end
+
+      # Check for any suspicious DNF
+      # Compute avg time per solve for the competitor
+      avg_per_solve = sum_of_times_for_rounds.to_f / completed_solves_for_rounds.size
+      # We want to issue a warning if the estimated time for all solves + DNFs goes roughly over the cumulative time limit by at least 10% (to reduce false positive).
+      if (number_of_dnf_solves + completed_solves_for_rounds.size) * avg_per_solve >= 1.1 * time_limit_for_round.centiseconds
+        @warnings << ValidationWarning.new(:results, competition_id,
+                                           SUSPICIOUS_DNF_WARNING,
+                                           round_ids: cumulative_round_ids.join(","),
+                                           person_name: person_name)
+      end
+    end
+
+    def results_similar_to(reference, reference_index, results)
+      # We do this programatically, but the original check_results.php used to do a big SQL query:
+      # https://github.com/thewca/worldcubeassociation.org/blob/b1ee87b318ff6e4f8658a711c19fd23a3ae51b9c/webroot/results/admin/check_results.php#L321-L353
+
+      similar_results = []
+      # Note that we don't want to treat a particular result as looking
+      # similar to itself, so we don't allow for results with matching ids.
+      # Further more, if a result A is similar to a result B, we don't want to
+      # return both (A, B) and (B, A) as matching pairs, it's sufficient to just
+      # return (A, B), which is why we require Result.id < h.resultId.
+      results.each_with_index do |r, index|
+        next if index >= reference_index
+        reference_solve_times = reference.solve_times
+        # We attribute 1 point for each similar solve_time, we then just have to count the points.
+        score = r.solve_times.each_with_index.count do |solve_time, solve_time_index|
+          solve_time.complete? && solve_time == reference_solve_times[solve_time_index]
+        end
+        # We have at least 3 matching values, consider this similar
+        if score > 2
+          similar_results << r
+        end
+      end
+      similar_results
+    end
+
+    def get_round_info(competition_rounds, round_id)
+      # This tries to get the round information from all the rounds in the competition.
+      round_info = competition_rounds[round_id]
+      unless round_info
+        # There is a legitimate situation where a round_id may be missing in the
+        # competition rounds: if it was a combined round and everyone made the cutoff!
```

Comment:
Let's take the example of a competition which has announced a combined final for Pyraminx with a 30s cutoff.
In the competition's rounds data there will be one round for pyraminx with `round_type_id` equal to `c`.
Then the competition happens, all competitors meet the cutoff (or the cutoff is removed, either leads to the same situation in the results).
When exporting through cubecomps/WA the `roundTypeId` will be `f` because according to the results the round is not combined.

We had a pretty long talk about that with the WRT and they were in favor of silently "fix" the issue by taking the information from the equivalent round with the cutoff in the competition's data (which also allows us to check with the announced cutoff that everybody did meet the cutoff).
In that case, in the `get_round_info` function, `round_id` would be `pyram-f`, and `equivalent_round_id` would be `pyram-c`, which would indeed be present in the competition's data.

Note that it doesn't work in the other way around: if it's a combined round in the results, then we absolutely need the correct round information in the competition's data!

Reply:
Thanks for the explanation! So the point I was missing is that cubecomps/WA determines the round type instead of taking the expected one.

Reply:
It would probably be helpful to add link to your comment in the *# comment*.

Reply:
Yes, and from the WRT point of view cubecomps/WA is putting the correct round type in there :)